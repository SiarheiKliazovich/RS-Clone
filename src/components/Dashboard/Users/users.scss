@import "../../../sass/variables";
@import "../../../sass/mixins";

* > {
    box-sizing: border-box;
}

.users__wrapper {
    width: 500px;
    margin: 0 auto;
    display: flex;
    flex-direction: column;
    align-items: center;
    background-color: $color-element-background-light;
}

.users__header {
    margin-bottom: 20px;
}

.users__header-button {
    margin-top: 30px;
    margin-bottom: 15px;
    width: 110px;
    height: 45px;
    color: $font-color-white;
    border: none;
    border-radius: 5px;
    background-color: $color-accent;
    font-size: 20px;
    cursor: pointer;

    &_active {
       background-color: $color-accent-hover; 
    }

    &:first-child {
        margin-right: 20px;
    }

    &:hover,
    &:active {
        background-color: $color-accent-hover;
    }
}