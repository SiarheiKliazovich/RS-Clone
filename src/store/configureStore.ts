/* eslint-disable @typescript-eslint/ban-ts-comment */
import { applyMiddleware, createStore } from "redux";
import thunkMiddleware from "redux-thunk";

import loggerMiddleware from "./logger";
import rootReducer from "../reducers";

const configureStore = () => {
  const middlewares = [loggerMiddleware, thunkMiddleware];
  const middlewareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(rootReducer, middlewareEnhancer);

  // @ts-ignore:next-line
  if (process.env.NODE_ENV !== "production" && module.hot) {
    // @ts-ignore:next-line
    module.hot.accept("./../reducers", () => store.replaceReducer(rootReducer));
  }

  return store;
};

export default configureStore;
