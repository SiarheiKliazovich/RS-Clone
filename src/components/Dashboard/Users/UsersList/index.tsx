import { FunctionComponent } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useTranslation } from "react-i18next";

import {
  deleteUser,
  getUsers,
  updateUser,
} from "../../../../services/apiUsers";
import {
  userDeleted,
  usersFetching,
  usersFetched,
  usersFetchingError,
} from "../../../../actions";
import { lightTheme, nightTheme } from "../../../../data/constants";
import { IStore } from "../../../../interfaces/store";
import { IAuthReducer } from "../../../../interfaces/authReducer";
import { IUsersListComponent } from "../../../../interfaces/usersListComponent";
import Hover from "../../../Hover";

import "./usersList.scss";

const UsersList: FunctionComponent<IUsersListComponent> = ({
  users,
  activeButton,
}) => {
  const appInterfaceStore = useSelector((state: IStore) => state.appInterface);
  const { isNightMode, isNavbarNightMode, accentColor } = appInterfaceStore;
  const authStore = useSelector((state: IAuthReducer) => state.auth);
  const { user } = authStore;

  const backgroundColor = isNightMode
    ? nightTheme.background.element
    : lightTheme.background.element;

  const dispatch = useDispatch();
  const role = activeButton === "1" ? "USER" : "SELLER";

  const { t } = useTranslation("dataLang");

  const deleteItem = async (id: string) => {
    try {
      await deleteUser(id);
      dispatch(userDeleted(id));
      dispatch(usersFetching());
      const users = await getUsers(role);
      dispatch(usersFetched(users));
    } catch {
      dispatch(usersFetchingError());
    }
  };

  const updateItem = async (id: string) => {
    try {
      await updateUser(id, { role: "MANAGER" });
      dispatch(usersFetching());
      const users = await getUsers(role);
      dispatch(usersFetched(users));
    } catch {
      dispatch(usersFetchingError());
    }
  };

  return (
    <>
      <div className="users-list__wrapper">
        <div className="users-list__item">
          <table>
            <tbody>
              <tr className="table-header">
                <th>#</th>
                <th>{t("usersTable.email")}</th>
                <th>{t("usersTable.id")}</th>
                {user.role === "ADMIN" ? (
                  <>
                    <th className="image-column"></th>
                    <th className="table-promotion"></th>
                  </>
                ) : null}
              </tr>
              {users.map((item, i) => {
                return (
                  <tr key={item._id}>
                    <th>{i + 1}</th>
                    <th>{item.email}</th>
                    <th>{item._id.slice(0, 8).toLocaleUpperCase()}</th>
                    {user.role === "ADMIN" ? (
                      <>
                        <th>
                          <button
                            className="button__delete-order"
                            onClick={() => deleteItem(item._id)}
                          >
                            <i className="fa fa-remove"></i>
                            {t("ordersTable.delete")}
                          </button>
                        </th>
                        <th className="table-promotion">
                          <Hover>
                            <button
                              style={{
                                backgroundColor: isNavbarNightMode
                                  ? backgroundColor
                                  : accentColor.static,
                              }}
                              className="button__promo"
                              onClick={() => updateItem(item._id)}
                            >
                              {t("usersTable.promote")}
                            </button>
                          </Hover>
                        </th>
                      </>
                    ) : null}
                  </tr>
                );
              })}
            </tbody>
          </table>
        </div>
      </div>
    </>
  );
};

export default UsersList;
