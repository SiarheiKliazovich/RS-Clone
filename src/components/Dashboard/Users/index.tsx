import { FunctionComponent, useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";

import { getUsers } from "../../../services/apiUsers";
import {
  usersFetching,
  usersFetched,
  usersFetchingError,
} from "../../../actions";

import { IUser } from "../../../interfaces/user";
import { IUsersList } from "../../../interfaces/usersList";
import { IUsersReducer } from "../../../interfaces/usersReducer";
import UsersList from "./UsersList";
import Spinner from "../../Spinner";

import "./users.scss";

const Users: FunctionComponent = () => {
  const [activeButton, setActiveButton] = useState<string | undefined>("1");
  const usersLoadingStatus = useSelector(
    (state: IUsersReducer) => state.usersReducer.usersLoadingStatus
  );

  const dispatch = useDispatch();

  const getUsersList = async (role: string) => {
    try {
      dispatch(usersFetching());
      const users = await getUsers(role);
      dispatch(usersFetched(users));
    } catch {
      dispatch(usersFetchingError());
    }
  };

  useEffect(() => {
    getUsersList("USER");
  }, []);

  const toggleHandler = (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    const id = e.currentTarget.dataset.id;
    setActiveButton(id);
  };

  const spinner = usersLoadingStatus === "loading" ? <Spinner /> : null;
  const usersList = usersLoadingStatus === "idle" ? <UsersList /> : null;

  return (
    <div className="users__wrapper">
      <div className="users__header">
        <button
          className={`users__header-button ${
            activeButton == "1" ? "users__header-button_active" : ""
          }`}
          data-id="1"
          onClick={(e) => {
            getUsersList("USER");
            toggleHandler(e);
          }}
        >
          Users
        </button>
        <button
          className={`users__header-button ${
            activeButton == "2" ? "users__header-button_active" : ""
          }`}
          data-id="2"
          onClick={(e) => {
            getUsersList("SELLER");
            toggleHandler(e);
          }}
        >
          Sellers
        </button>
      </div>
      {spinner}
      {usersList}
    </div>
  );
};
export default Users;
