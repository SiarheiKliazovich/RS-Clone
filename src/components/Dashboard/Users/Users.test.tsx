import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import Users from ".";

const mockStore = configureStore([]);

describe("Users", () => {
  // eslint-disable-next-line
  let store: any;

  beforeEach(() => {
    store = mockStore({
      appInterface: {
        accentColor: {
          static: "#000",
          hover: "#fff",
        },
      },
      usersReducer: {
        usersLoadingStatus: "idle",
        users: [],
      },
      auth: {
        user: {
          role: "USER",
        },
      },
    });
  });

  test("renders correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <Users />
        </Router>
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
