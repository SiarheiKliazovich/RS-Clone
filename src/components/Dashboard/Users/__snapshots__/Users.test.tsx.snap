// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users renders correctly 1`] = `
<div
  class="users__wrapper"
  style="background-color: rgb(255, 255, 255);"
>
  <div
    class="users__header"
  >
    <button
      class="users__header-button users__header-button_active"
      data-id="1"
      style="background-color: rgb(255, 255, 255);"
    >
      buttonsUsersList.users
    </button>
    <button
      class="users__header-button "
      data-id="2"
      style="background-color: rgb(0, 0, 0);"
    >
      buttonsUsersList.sellers
    </button>
  </div>
  <div
    class="users-list__wrapper"
  >
    <div
      class="users-list__item"
    >
      <table>
        <tbody>
          <tr
            class="table-header"
          >
            <th>
              #
            </th>
            <th>
              usersTable.email
            </th>
            <th>
              usersTable.id
            </th>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
